bin/src/hickory-dns.rs

```diff
@@ -224,10 +230,22 @@ fn run() -> Result<(), String> {
 
     #[cfg(feature = "prometheus-metrics")]
     if !args.disable_prometheus && !config.disable_prometheus() {
-        let socket = args
+        let mut socket = args
             .prometheus_listen_addr
             .unwrap_or(config.prometheus_listen_addr());
 
+        // bind to select a port, then drop the port so it can be used
+        // needed for dynamic port selection (port 0) as it is currently not possible to
+        // obtain the selected port from PrometheusBuilder after it binds the SocketAddr
+        // during PrometheusBuilder::install() or PrometheusBuilder::build()
+        if socket.port() == 0 {
+            let listener = std::net::TcpListener::bind(socket)
+                .map_err(|e| format!("failed to bind prometheus metrics socket {e}"))?;
+            socket = listener.local_addr().map_err(|e| {
+                format!("failed to get local addr for prometheus metrics socket {e}")
+            })?;
+        }
```

Comment:
There's a chance this could lead to test flakes due to races to re-use dynamically-selected ports. Unfortunately, there's no API to provide a pre-bound TCP listener to the builder. If we run into issues with this, I think the next best option is to use `PrometheusHandle::render()` and run our own HTTP server.

Reply:
I see, and potentially running the http server directly would provide a lot more flexibility.

During creation of the workaround I've as well opened a ticket suggesting to allow creating the http listener with a pre-configured TcpListener. https://github.com/metrics-rs/metrics/issues/573

This would not solve topics like http2 support or TLS during scraping which could be handled when running the server directly.